Add vertical coordinate design document (#230)
E3SM-Project/Omega: components/omega/doc/design/VertCoord.md

(omega-design-vert-coord)=

> # Vertical Coordinate

Reviewer: One problem we've had in MPAS-Ocean that this document perpetuates is the arbitrary use of level vs. layer.  Do we want to settle on one of these, once an for all, and stick with it?  My (fairly strong) preference is for layer over level.

Reviewer 2: I also prefer layer

Author: I agree we should settle on something consistent and I also would prefer layer.

Reviewer 3: I agree as well.

## 1 Overview
The vertical coordinate module will be responsible for computing and storing information related to the vertical mesh in Omega.

> Since Omega will be a non-Boussinesq model, the vertical independent variable will be pressure- as opposed to $z$-based.

Reviewer: ```suggestion
Since Omega will be a non-Boussinesq model, the vertical independent variable will be pressure-based (expressed in terms of the pseudo-height) as opposed to $z$-based.
```

Similar to MPAS-Ocean's support for tilted height ($z^\star$) coordinates, Omega V1 will support a general vertical coordinate, where pseudo thickness $h$ varies in proportion to the total pseudo thickness of the water column. 
The prognostic variable in the layered continuity equation is pseudo thickness, which can be used to calculate the total pressure at a given vertical level.

> The vertical height of a given level is still necessary in the model to compute sea level, the geopotential, and derivatives with respect to $z$.

Reviewer: ```suggestion
The geometric height of a given level is still necessary in the model to compute sea level, the geopotential, and derivatives with respect to $z$.
```

> The height of a layer ($\Delta z$) can be found using the pseudo thickness, density, reference density, and known bottom elevation relative to the geoid (positive up).

Reviewer: ```suggestion
The geometric thickness of a layer ($\Delta z$) can be found using the pseudo thickness, density, reference density, and known bottom elevation (positive up) relative to the reference geoid.
```

Reviewer: "reference geoid" could also be "static geoid".

The vertical coordinate module will also serve as a container for information related to the variable number of active vertical levels for a given ocean column (due to variations in bottom elevation and ice shelf cavities).

## 2 Requirements

### 2.1 Requirement: Compute and store the total pressure at a given level

The total pressure at the bottom of each level will be computed and stored within the vertical coordinate module.
This involves a surface-down summation of the pseudo thickness variable times the gravitational acceleration, $g$, and the reference density, $\rho_0$, starting with the surface pressure.
The pressure variable will be needed in the computation of the TEOS-10 equation of state and the pressure gradient.
The variable will be registered with the `IOStreams` framework to enable it to be output during runtime.

### 2.2 Requirement: Compute and store the $z$ location of a given level

> The $z$ location at the top of each level will be computed and stored within the vertical coordinate module.

Reviewer: Again, could this be at layer interfaces rather than the top of each level?

This involves adding a vertically-weighted bottom pressure perturbation to the reference (initial) thickness.
This desired thickness will be computed and stored within the vertical coordinate module.
The vertical advection algorithm will use this desired thickness to compute the vertical transport.

### 2.6 Desired: General Arbitrary Lagrangian Eulerian (ALE) coordinate support

In future versions of Omega, the $p^\star$ coordinate will be extended to a more general ALE coordinate

### 2.7 Desired: Vertical Lagrangian remapping (VLR) support

The vertical coordinate module should retain flexibility to support VLR in future versions of Omega.

## 3 Algorithmic Formulation
In the layered non-Boussinesq equations solved in Omega (see [V0 governing equation document](OmegaV1GoverningEqns) for details), the prognostic variable for cell $i$ and layer $k$ is the pseudo thickness, $\tilde{h}_{i,k}$, so that the geometric thickness (in meters) is a diagnostic variable defined as:

$$ \Delta z_{i,k} = \rho_0 \alpha_{i,k} \tilde{h}_{i,k}. $$

The pressure at vertical cell interfaces is the found by summing the pseudo thicknesses:

> $$  p_{i,k+1/2} = p_{i}^{surf} + g\rho_0 \sum_{k^\prime=1}^k \tilde{h}_{i,k^\prime}. $$

Reviewer: Would it be truer to C++ if we were to use zero-based indexing for $k$?

Author: Possibly, but I think it's a bit awkward. I don't often see equation summations use zero-based indexing.

Reviewer: Okay, no problem.  Let's stick with what you have.

and at the midpoint by

$$ p_{i,k} = p_{i}^{surf} + g\rho_0 \sum_{k'=1}^{k-1} \tilde{h}_{i,k'} + \frac{1}{2} g\rho_0 \tilde{h}_{i,k} $$

The $z$ location of cell interfaces is found by summing the pseudo thicknesses from the bottom multiplied by the specific volume:

> $$ z_{i,k}^{top} = z_i^{floor} + \rho_0 \sum_{k^\prime=k}^{K_{max}} \alpha_{i,k^\prime} \tilde{h}_{i,k^\prime}, $$

Reviewer: Shouldn't this be $z_{i,k+1/2}$, similar to $p_{i,k+1/2}$ above?

Author: yes, thanks!

where $z_i^{floor}$ is the (positive-up) bottom elevation.
The $z$ location of cell centers is given by: 

$$ z_{i,k} = z_i^{floor} + \frac{1}{2} \rho_0\alpha_{i,k} \tilde{h}_{i,k} + \rho_0\sum_{k^\prime= k+1}^{K_{max}} \alpha_{i,k^\prime} \tilde{h}_{i,k^\prime}. $$

Initially, the geopotential is the sum of the $z$ height times $g$. 
In the future, it will include contributions from the tidal potential ($\Phi_{tp}$) and self attraction and loading ($\Phi_{SAL}$):

$$ \Phi_{i,k} = \left( gz_{i,k} + \Phi_{tp} + \Phi_{SAL} \right). $$

The desired pseudo thickness used for the $p^\star$ coordinate is:

> $$\tilde{h}_k^{p^\star} = \tilde{h}_k^{rest} + \tilde{h}_k^{p}, $$

Reviewer: Could we call this $ref$ or similar instead of $rest$?  In MPAS-Ocean, the use of a "resting thickness" has caused a lot of confusion, especially in ice-shelf cavities, because it is missing the surface pressure.  It seems like the reference pseudo-thicknesses wouldn't have this particular issue but I also don't think there is a strict requirement that the reference pseudo-thicknesses be consistent with the model at rest.

where the pressure alteration, $\tilde{h}_k^{p}$, is given by:

> $$ \tilde{h}_k^{p} =\frac{(p_B-p_{surf})}{g\rho_0} \frac{W_k\tilde{h}_k^{rest}}{\sum_{k^\prime=1}^K W_{k^\prime}\tilde{h}_{k^\prime}^{rest}}.$$

Reviewer: The weights $W_k$ should be explained.

Reviewer: I don't think this equation and the previous one are consistent.  This seems like the full pseudo-thickness that each layer should have, not just the perturbation from the resting thickness.  I would benefit from a discussion of this if I'm missing something.

Reviewer 2: I see. Sure, that sounds good.

## 4 Design

### 4.1 Data types and parameters

The `VertCoord` class will be used to compute pressure, $z$ height, and geopotential at each layer.
It will also contain the vertical loop bounds for cells, edges, and vertices.
```c++
class VertCoord {
    public:
        I4 NVertLevels;
        I4 NVertLevelsP1;

        // Variables computed 
        Array2DReal PressureInterface;
        Array2DReal PressureMid;        

>         Array2DReal ZInterface;

Reviewer: Big thumbs up to `PressureInterface` and `ZInterface` (as opposed to bottom or top, as described above).

        Array2DReal ZMid;
        Array2DReal GeopotentialMid;
        Array2DReal LayerThicknessPStar;

        // Vertical loop bounds
        Array1DI4 MinLevelCell;
        Array1DI4 MaxLevelCell;
        Array1DI4 MinLevelEdgeTop;
        Array1DI4 MaxLevelEdgeTop;
        Array1DI4 MinLevelEdgeBot;     
        Array1DI4 MaxLevelEdgeBot;
        Array1DI4 MinLevelVertexTop;
        Array1DI4 MaxLevelVertexTop;     
        Array1DI4 MinLevelVertexBot;
        Array1DI4 MaxLevelVertexBot;

        // p star coordinate variables
        Array2DReal VertCoordMovementWeights;
        Array2DReal RefLayerThickness;

        // Variables from HorzMesh
        Array2DReal BottomDepth;
    private:

        // Variables from HorzMesh
        I4 NCells;
        I4 NEdges;
        I4 NVertices;
        Array2DReal CellsOnEdge;
        Array2DReal CellsOnVertex;

        static VertCoord *DefaultVertCoord;
        static std::map<std::string, std::unique_ptr<VertCoord>> AllVertCoords;
}
```

### 4.2 Methods

```c++
class VertCoord{
    public:
        static VertCoord *create();
        void computePressure();
        void computeZHeight();
        void computeGeopotential();

>         void computePStarThickness();

Reviewer 2: Should we consider generalizing this to `computeTargetPseudoThickness` or something along those lines that's more succinct?

Author: That sounds good to me. Maybe just `computeTargetThickness` to cut down on length and since we don't include `Pseudo` in any other variable names (e.g. `layerThickness`)?

In the model, $z$ is defined as being positive up from the geoid.
The $z$ location is calculated in a bottom-up summation of the pseudo thickness times the specific volume and reference density, starting with the known bottom elevation.
The $z$ location of the interfaces is needed to compute the geopotential gradient and the sea level.
The variable will be registered with the `IOStreams` framework to enable it to be output during runtime.

### 2.3 Requirement: Compute and store the vertical layer bounds for each horizontal cell

All tendency terms require a variable vertical loop range to account for variation in bottom elevation and ice shelf cavities.
These bounds will be computed on construction in the vertical coordinate module and used for setting the loop bounds for active layers in calculations over the vertical.

### 2.4 Requirement: Compute and store the geopotential

The geopotential will be computed and stored within the vertical coordinate module.
Initially this will be calculated by summing together the $z$ location times $g$.
In the future, tidal potential and self attraction and loading will also have optional (default off) contributions to the geopotential.

> ### 2.5 Requirement: Compute and store the desired thickness used in the $p^\star$ vertical coordinate

Reviewer 2: The current proposal in the design doc, as I understand it, assumes a time-constant $\tilde h_{i,k}^{ref}$ that is defined assuming $\rho = \rho_0$, and is thus equal to $h_{i,k}^{ref}$. This is the field that is then scaled by the column-integrated pressure deviation and $W_k$.

My concern with this approach is that it may make having a coordinate that moves with presure and density surfaces more difficult.

Suppose we want a hybrid coordinate that is approximately z-level at the surface and approximately isopycnal below the mixed layer. With the current proposal, we'd likely take an initial ocean density field and specify the $h_{i,k}^{ref}$ values to reflect the hybrid coordinate at the start of the simulation. 

Now suppose the pycnocline is displaced downward due to enhanced entrainment and the column-integrated pressure is similar. I think the only way to get $\tilde h_{i,k}^{p^\star}$ to follow the pycnocline would be to update $W_{i,k}$ with the dynamics.

My question is whether we could produce a better vertical coordinate with less effort (in terms of updating $W_{i,k}$) by taking $p_{i,k}^{ref}$ as the quantity that stays constant in time.

I think this means that the desired layer interfaces would be located at these pressure levels:
$$p_{i,k}^{new} = p_k^{ref} + (p_B-p_{surf} - \sum_{k^\prime=1}^K p_{k^\prime}^{ref}) \frac{W_k p_k^{ref}}{\sum_{k^\prime=1}^K W_{k^\prime} p_{k^\prime}^{ref}}$$

and the pseudo thickness would be computed after that as:
$$\tilde h_{i,k}^{p*} = \frac{1}{g \rho_0} (p_{i,k}^{new} - p_{i,k-1}^{new})$$

In the pycnocline deepening example, I think we'll get more movement with the pycnocline in this approach. My sense is that generally with this approach $p_k^{top}$ is more steady with time while $z_k^{top}$ varies more with time whereas before $p_k^{top}$ varied more with time while $z_k^{top}$ stayed more steady. I think the former is more desirable. Let me know if I've misstepped here or not explained myself clearly

(omega-design-vert-coord)=
# Vertical Coordinate

## 1 Overview
The vertical coordinate module will be responsible for computing and storing information related to the vertical mesh in Omega.

> Since Omega will be a non-Boussinesq model, the vertical independent variable will be pressure- as opposed to $z$-based.

Reviewer: @mark-petersen, I don't know that I entirely agree.  We can define $p$ at vertical interfaces as long as we know the surface and bottom pressure based on the definition of the pressure coordinate.  But, just as we have to compute $p$ in MPAS-Ocean using the equation of state, we have to compute $z$ in Omega using the equation of state.  This means that it is entirely possible depending on the order of operations that $p$ will be known sometimes in situations where $z$ is not (most notably in evaluating the equation of state).  So I would stand by the original wording or at least something closer to it.  $p$ and $z$ are not equally independent in my view.

> Since Omega will be a non-Boussinesq model, the vertical independent variable will be pressure- as opposed to $z$-based.

Author: I agree that while $z$ can be used to choose layer locations, but it is a function of density and therefore not independent.

> Similar to MPAS-Ocean's support for tilted height ($z^\star$) coordinates, Omega will support a general ($p^\star$) vertical coordinate.

Reviewer 4: We can actually do either, since there is an easy computation between $z$ and $p$ and both are available.

```suggestion
Similar to MPAS-Ocean, the vertical coordinate will be Arbitrary Lagrangian-Eularian (ALE). This means that the vertical layer interfaces will be computed within the ALE routine, and may be chosen to be z-star ($z^\star$, with layer thickness varying in proportional to sea surface height in $z$) or p-star ($p^\star$, varying in proportion to bottom pressure, in $p$). P-star is a more natural choice for a non-Boussinesq model.
```

Reviewer: > varies in proportion to the bottom pressure.

It that correct?  I thought it would vary proportion to the total mass-thickness of the water column.

The prognostic variable in the layered continuity equation is mass thickness, which can be used to calculate the total pressure at a given vertical level.

> The vertical height of a given level is still necessary in the model to compute sea level, the geopotential, and derivatives with respect to $z$ (vertical mixing, etc).

Reviewer 3: I don't think vertical mixing is a good example.  It doesn't really require z derivatives that's just what we are used to

The height of a layer ($\Delta z$) can be found using the mass thickness, density and known bottom elevation relative to the geoid (positive up).
The vertical coordinate module will also serve as a container for information related to the variable number of vertical levels for a given ocean column due to variations in bottom elevation and ice shelf cavities.

## 2 Requirements

### 2.1 Requirement: Compute and store the total pressure at a given level

> The total pressure at the bottom of each level will be computed and stored within the vertical coordinate module.

Author: I agree this is something we should consider.

> The total pressure at the bottom of each level will be computed and stored within the vertical coordinate module.

Reviewer: Another consideration.  I would prefer to have the pressure at each layer interface (there is one more layer interface than the number of layers).  This makes various computations (e.g. averaging or differencing) easier and it also doesn't arbitrarily favor the bottom vs. the top of a given layer.  In Compass and Polaris, we have used `z_interface` pretty extensively in this way.

I think we absolute *have* to have the pressure at interfaces (or bottoms of layers) because it is too difficult to reconstruct this from the pressure at the middle of layers.  But we could consider always averaging to get the pressure at the middle of layers.  If we do this, the best way to avoid having to treat the top and bottom of each layer as a special case would be to have pressure at all layer interfaces (including the top of the ocean).

> The total pressure at the bottom of each level will be computed and stored within the vertical coordinate module.

Reviewer: Another point, for the higher-order pressure gradient, it isn't obvious that we would want to compute the equation of state at the middle of layers as opposed to at quadrature points.  So that may be another reason to prefer to just store it at the points where it is defined (layer interfaces).

> The total pressure at the bottom of each level will be computed and stored within the vertical coordinate module.

Author: I makes sense to me to primarily store pressure and z at interfaces. We could compute the mid-layer values as needed for the centered pressure gradient and diagnostic output. 

@katsmith133, are `ZMid` and `PressureMid` needed for vertical mixing?

All tendency terms require a variable vertical loop range to account for variation in bottom elevation and ice shelf cavities.
These bounds will be computed in the vertical coordinate module and used for setting masking arrays that will be used to determine where calculations are valid in the vertical.

### 2.4 Requirement: Compute and store the geopotential

The geopotential will be computed and stored within the vertical coordinate module.
Initially this will be calculated by summing together the $z$ location times $g$.
In the future, tidal potential and self attraction and loading will also have contributions to the geopotential.

### 2.5 Requirement: Compute and store the desired thickness used in the $p^\star$ vertical coordinate

The $p^\star$ vertical coordinate specifies a desired layer thickness that stretches the initial thicknesses by the bottom pressure anomaly.
This involves adding a vertically-weighted bottom pressure perturbation to the a reference (initial) thickness.
This desired thickness will be computed and stored within the vertical coordinate module.
The vertical advection algorithm will use this desired thickness to compute the vertical transport.

> ### Desired: General Arbitrary Lagrangian Eulerian (ALE) coodrinate support

Reviewer 4: ```suggestion
### Desired: General Arbitrary Lagrangian Eulerian (ALE) coodinate support
```

In future versions of Omega, the $p^\star$ coordinate will be extended to a more general ALE coordinate

### Desired: Vertical Lagrangian remapping (VLR) support

The vertical coordinate module should retain flexibility to support VLR in future versions of Omega.

## 3 Algorithmic Formulation
The non-Boussinesq momentum equation is

$$ 
\frac{D \mathbf{u}_h}{D t } + f\boldsymbol{k}\times \mathbf{u}_h + \left(v\nabla_A p + \nabla_A \Phi \right) = \boldsymbol{\mathcal{F}}. 

> $$

Author: I agree. Have we settled on a form of the momentum equation that we're happy with?

> $$

Author: I linked to the V1 equations document instead of re-writing the momentum equation here. I tried to make all other notation consistent.

$$ z_{i,k+1/2} = z_i^{floor} + \sum_{k=k^\prime}^{N} v_{i,k^\prime} h_{i,k^\prime}, $$

where $z_i^{floor}$ is the (positive-up) bottom elevation.
The $z$ location of cell centers is given by: 

$$ z_{i,k} = z_i^{floor} + \frac{1}{2} v_{i,k} h_{i,k} + \sum_{k^\prime= k+1}^{N} v_{i,k^\prime} h_{i,k^\prime}. $$

Initially, the geopotential is the sum of the $z$ height times $g$. 
In the future, it will include contributions from the tidal potential ($\Phi_{tp}$) and self attraction and loading ($\Phi_{SAL}$):

$$ \Phi_{i,k} = \left( gz_{i,k} + \Phi_{tp} + \Phi_{SAL} \right). $$

The desired mass thickness used for the $p^\star$ coordinate is:

$$h_k^{p^\star} = h_k^{rest} + h_k^{p}, $$

where the the pressure alteration, $h_k^{p}$, is given by:

$$ h_k^{p} =\frac{(p_B-p_{surf})}{g} \frac{W_kh_k^{rest}}{\sum_{k^\prime=1}^K W_{k^\prime}h_{k^\prime}^{rest}}.$$

## 4 Design

### 4.1 Data types and parameters

The `VertCoord` class will be used to compute pressure, $z$ height, and geopotential at each layer.
It will also contain the vertical loop bounds for cells, edges, and vertices.
```c++
class VertCoord {

>     public:

Reviewer 2: Is there an advantage to having `BottomDepth` in `VertCoord`? I presume $$z_i^{floor} = \text{BottomDepth} $$

>     public:

Reviewer 2: I meant to propose making `BottomDepth` private rather than public along with the rest of the horizontal mesh variables, not remove it altogether. Now realizing I didn't quite communicate that.

>     public:

Reviewer 2: I could see arguing either way. `VertCoord` is mostly mutable and `HorzMesh` is mostly immutable, right? With the exception of `VertCoordMovementWeights` and `RefLayerThickness`. Maybe I do like it in `VertCoord` better and I'm just too accustomed to MPAS-O.

>     public:

Author: It seems like having a vertical mesh stream would be a good option.

    private:
        I4 NVertLevels;
        I4 NVertLevelsP1;

> 

Author: Thanks for noticing this. I figure since we have the option to have multiple horizontal meshes, we should support multiple vertical coordinates too. Let me know if you think otherwise.

        // Variables computed 
        Array2DReal PressureInterface;
        Array2DReal PressureMid;        
        Array2DReal ZInterface;
        Array2DReal ZMid;
        Array2DReal GeopotentialMid;
        Array2DReal LayerThicknessPStar;

        // Vertical loop bounds
        Array2DI4 MinLevelCell;
        Array2DI4 MaxLevelCell;
        Array2DI4 MinLevelEdgeTop;
        Array2DI4 MaxLevelEdgeTop;
        Array2DI4 MinLevelEdgeBot;     
        Array2DI4 MaxLevelEdgeBot;
        Array2DI4 MinLevelVertexTop;
        Array2DI4 MaxLevelVertexTop;     
        Array2DI4 MinLevelVertexBot;
        Array2DI4 MaxLevelVertexBot;

        // p star coordinate variables
        Array2DReal VertCoordMovementWeights;
        Array2DReal RefLayerThickness;

        // Variables from HorzMesh
        I4 NCells;
        I4 NEdges;
        I4 NVertices;
        Array2DReal CellsOnEdge;
        Array2DReal CellsOnVertex;
        Array2DReal BottomDepth;
}
```
The user will select vertical coordinate options at runtime in the input yaml file under the vertical coordinate section:
```yaml

>     VerticalCoordinate:

Reviewer 5: Getting NVertLevels from the Config was really just meant to be a temporary measure while you were building IOStreams and revamping IO. I agree it should just come from the mesh file by default.

>     VerticalCoordinate:

Author: Yes, I agree it should come from the mesh.

```c++
void VertCoord::computePressure(const Array2DReal &PressureInterface,
                                const Array2DReal &PressureMid,
                                const Array2DReal &LayerThickness,
                                const Array2DReal &SurfacePressure) {

}                                
```

The public `computeZHeight` will sum the mass thicknesses times $v$ from the bottom later up, starting with the bottom elevation
This will be done with a `parallel_scan` inside a `parallal_for over the mesh cells using hierarchical parallelism.
```c++
void VertCoord::computeZHeight(const Array2DReal &ZInterface,
                               const Array2DReal &ZMid,
                               const Array2DReal &LayerThickness,
                               const Array2DReal &SpecVol,
                               const Array2DReal &BottomDepth) {

}    
```

The public `computeGeopotential` will sum together the $z$ height times $g$, the tidal potential, and self attraction and loading:
```c++
void VertCoord::computeGeopotential(const Array2DReal &GeopotentialMid,
                                    const Array2DReal &ZMid,

>                                     const Array2DReal &TidalPotential

Author: For now, I've just added: 
```
Tidal potential forcing and self attraction and loading will de default-off features.
The will be added to (or excluded from) the geopotential based on config flags
```

                                const Array2DI4 &MinLevelCell,
                                const Array2DI4 &MaxLevelCell,
                                const Array2DI4 &CellsOnEdge) {

}
```

The private `minMaxLevelVertex` will determine the various vertical loop bounds on vertices:
```c++
void VertCoord::minMaxLevelVertex(const Array2DI4 &MinLevelVertexTop,
                                  const Array2DI4 &MaxLevelVertexTop,
                                  const Array2DI4 &MinLevelVertexBot,
                                  const Array2DI4 &MaxLevelVertexBot,
                                  const Array2DI4 &MinLevelCell,
                                  const Array2DI4 &MaxLevelCell,
                                  const Array2DI4 &CellsOnVertex) {
                                  
} 
```

#### 4.2.5 Destruction and removal

No operations are needed in the destructor.
The erase method will remove a named vertical coordinate instance, whereas the clear method will remove all of
them. 
Both will call the destructor in the process.
```c++
void VertCoord::erase(const std::string &Name);
void VertCoord::clear();
```

## Verification and Testing

### Test: 

> Unit tests will be used to test each of the computations (computePressure, computeZHeight, computeGeopotential, computePStarThickness) for a given mass thickness array. Comparison will be made to a ``truth'' vertical mesh.

Reviewer 2: Should we also test these routines with a z-level coordinate to make sure bounds and weighting work with spatially variable `minLevelCell` and `maxLevelCell`?

(omega-design-vert-coord)=
# Vertical Coordinate

## 1 Overview
The vertical coordinate module will be responsible for computing and storing information related to the vertical mesh in Omega.
Since Omega will be a non-Boussinesq model, the vertical independent variable will be pressure- as opposed to $z$-based.
Similar to MPAS-Ocean's support for tilted height ($z^\star$) coordinates, Omega V1 will support a general ($p^\star$) vertical coordinate, where mass thickness $h$ varies in proportion to the total mass thickness of the water column. 
The prognostic variable in the layered continuity equation is mass thickness, which can be used to calculate the total pressure at a given vertical level.
The vertical height of a given level is still necessary in the model to compute sea level, the geopotential, and derivatives with respect to $z$ (vertical mixing, etc).
The height of a layer ($\Delta z$) can be found using the mass thickness, density and known bottom elevation relative to the geoid (positive up).

> The vertical coordinate module will also serve as a container for information related to the variable number of vertical levels for a given ocean column due to variations in bottom elevation and ice shelf cavities.

Reviewer 6: I know I was on leave, so maybe I missed a decision somewhere, but technically its the active number of layers that is variable, correct? There will still be the same number of vertical levels for every column, just the number of layers we loop over (or mask) will change based upon bottom elevation and ice shelf cavities? 

Maybe this is nit-picky, but maybe something like `variable number of active vertical levels` or whatever terminology we want to use to differentiate between `NVertLevels` and the quantity that is `maxLevelCell - minLevelCell`.

This involves a surface-down summation of the mass thickness variable times the gravitational acceleration, $g$, starting with the surface pressure.
The pressure variable will be needed in the computation of the TEOS-10 equation of state and the pressure gradient.
The variable will be registered with the `IOStreams` framework to enable it to be output during runtime.

### 2.2 Requirement: Compute and store the $z$ location of a given level

The $z$ location at the top of each level will be computed and stored within the vertical coordinate module.
In the model, $z$ is defined as being positive up from the geoid.
The $z$ location is calculated in a bottom-up summation of the mass thickness times the specific volume, starting with the known bottom elevation.
The $z$ location of the interfaces is needed to compute the geopotential gradient and the sea level.
The variable will be registered with the `IOStreams` framework to enable it to be output during runtime.

### 2.3 Requirement: Compute and store the vertical level bounds for each horizontal cell

All tendency terms require a variable vertical loop range to account for variation in bottom elevation and ice shelf cavities.

> These bounds will be computed in the vertical coordinate module and used for setting masking arrays that will be used to determine where calculations are valid in the vertical.

Reviewer 6: From our last meeting, sounds like this will be loop bounds instead of masking?

Author: yes, I'll update that.

### 2.4 Requirement: Compute and store the geopotential

The geopotential will be computed and stored within the vertical coordinate module.
Initially this will be calculated by summing together the $z$ location times $g$.

> In the future, tidal potential and self attraction and loading will also have contributions to the geopotential.

Reviewer 6: `...will also be options to contribute to the geopotential.` or `...will also have (default off) contributions to the geopotential`?

Author: This is a good clarification. Fixed.

### 2.5 Requirement: Compute and store the desired thickness used in the $p^\star$ vertical coordinate

The $p^\star$ vertical coordinate specifies a desired layer thickness that stretches the initial thicknesses by the bottom pressure anomaly.
This involves adding a vertically-weighted bottom pressure perturbation to the reference (initial) thickness.
This desired thickness will be computed and stored within the vertical coordinate module.
The vertical advection algorithm will use this desired thickness to compute the vertical transport.

> ### Desired: General Arbitrary Lagrangian Eulerian (ALE) coordinate support

Reviewer 6: Maybe add numbers to these too? The way it looks in the compiled version it kind of looks like these are a new section instead of just a part of Section 2.

Author: yes, thanks for pointing that out.

In future versions of Omega, the $p^\star$ coordinate will be extended to a more general ALE coordinate

### Desired: Vertical Lagrangian remapping (VLR) support

The vertical coordinate module should retain flexibility to support VLR in future versions of Omega.

## 3 Algorithmic Formulation
The non-Boussinesq momentum equation is

$$ 
\frac{D \mathbf{u}_h}{D t } + f\boldsymbol{k}\times \mathbf{u}_h + \left(\alpha\nabla_A p + \nabla_A \Phi \right) = \boldsymbol{\mathcal{F}},
$$

> where $\mathbf{u}_h$ is the horizontal velocity, $f$ is the Coriolis parameter, $\alpha = \frac{1}{\rho}$ is the specific volume, $\rho = \rho(\Theta,S,p)$ is the density, $p$ is the hydrostatic pressure, $\Phi$ is the geopotential, and $\boldsymbol{\mathcal{F}}$ are the dissipative terms.

Reviewer 6: Alpha is not rendering correctly here, maybe try this?
```suggestion
where $\mathbf{u}_h$ is the horizontal velocity, $f$ is the Coriolis parameter, $\alpha$ = $\frac{1}{\rho}$ is the specific volume, $\rho = \rho(\Theta,S,p)$ is the density, $p$ is the hydrostatic pressure, $\Phi$ is the geopotential, and $\boldsymbol{\mathcal{F}}$ are the dissipative terms.
```

Author: good catch!

```c++
VertCoord *VertCoord::create(const std::string &Name,
                             const HorzMesh *Mesh,
                             int NVertLevels,
                             Config *Options);
```

#### 4.2.2 Initialization

The init method will create the default vertical coordinate and return an error code:
```c++
int VertCoord::init();
```

#### 4.2.3 Retrieval

There will be methods for getting the default and non-default vertical coordinate instances:
```c++
VertCoord *VertCoord::getDefault();
VertCoord *VertCoord::get(const std::string &Name);
```

#### 4.2.4 Computation

The public `computePressure` will sum the mass thicknesses times $g$ from the top layer down, starting with the surface pressure.
This will be done with a `parallel_scan` inside a `parallel_for` over the mesh cells using hierarchical parallelism.
```c++
void VertCoord::computePressure(const Array2DReal &PressureInterface,
                                const Array2DReal &PressureMid,
                                const Array2DReal &LayerThickness,
                                const Array2DReal &SurfacePressure) {

}                                
```

> The public `computeZHeight` will sum the mass thicknesses times $\alpha$ from the bottom later up, starting with the bottom elevation

Reviewer 6: ```suggestion
The public `computeZHeight` will sum the mass thicknesses times $\alpha$ from the bottom later up, starting with the bottom elevation.
```

Author: fixed, thanks!

> This will be done with a `parallel_scan` inside a `parallal_for over the mesh cells using hierarchical parallelism.

Reviewer 6: ```suggestion
This will be done with a `parallel_scan` inside a `parallal_for` over the mesh cells using hierarchical parallelism.
```

Author: fixed, thanks!

}    
```

The public `computeGeopotential` will sum together the $z$ height times $g$, the tidal potential, and self attraction and loading:
```c++
void VertCoord::computeGeopotential(const Array2DReal &GeopotentialMid,
                                    const Array2DReal &ZMid,
                                    const Array2DReal &TidalPotential
                                    const Array2DReal &SelfAttractionLoading) {
                                  
} 
```
Tidal potential forcing and self attraction and loading will be default-off features.
The will be added to (or excluded from) the geopotential based on config flags.

The public `computePStarThickness` will determine the desired mass thickness used for the $p^\star$ vertical coordinate:
```c++
void VertCoord::computePStarThickness(const Array2DReal &LayerThicknessPStar,
                                      const Array2DReal &VertCoordMovementWeights,
                                      const Array2DReal &RefLayerThickness) {
                                  
} 
```

The private `minMaxLevelEdge` will determine the various vertical loop bounds on edges:
```c++

> void VertCoord::minMaxLevelEdge(const Array2DI4 &MinLevelEdgeTop,

Reviewer 6: Maybe we can just combine these into one called `minMaxLevel` that provides all of them? Unless there is an advantage to breaking it up?

Author: I think that makes sense.
